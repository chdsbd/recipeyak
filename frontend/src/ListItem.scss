.listitem-button-container {
  display: flex;
  justify-content: space-between;
}

.listitem-text {
  white-space: pre-wrap;
}

.cursor--pointer {
  cursor: pointer;
  // need to override all of the default cursors that come with the default
  // user stylesheet
  & > * {
    cursor: pointer;
  }
}

.unsaved-changes {
  display: flex;
  justify-content: space-between;
  align-items: center;
}

.is-italic {
  font-style: italic;
}
