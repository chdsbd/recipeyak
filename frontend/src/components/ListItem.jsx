import React from 'react'

class ListItem extends React.Component {
  constructor (props) {
    super(props)
    this.state = {
      text: props.text || '',
      editing: false,
      unsavedChanges: false
    }
  }

  componentWillMount () {
    document.addEventListener('mousedown', this.handleGeneralClick)
  }

  componentWillUnmount () {
    document.removeEventListener('mousedown', this.handleGeneralClick)
  }

  // ensures that the list item closes when the user clicks outside of the item
  handleGeneralClick = e => {
    const clickedInComponent = this.element && this.element.contains(e.target)
    if (clickedInComponent) return
    this.setState((prevState, props) => ({
      editing: false,
      unsavedChanges: (prevState.editing && prevState.text !== props.text) || prevState.unsavedChanges
    }))
  }

  handleInputChange = e => {
    this.setState({ [e.target.name]: e.target.value })
  }

  enableEditing = () => {
    this.setState({
      editing: true,
      unsavedChanges: false
    })
  }

  discardChanges = () => {
    this.setState((prevState, props) => ({
      editing: false,
      text: props.text,
      unsavedChanges: false
    }))
  }

  handleFocus = e => {
    e.target.select()
  }

  add () {
    console.log(this.state)
    // TODO: update store
  }

  cancel = e => {
    e.stopPropagation()
    this.setState((prevState, props) => ({
      editing: false,
      text: props.text
    }))
  }

  update = async e => {
    e.stopPropagation()
    // if the text is empty, we should just delete the item instead of updating
    if (this.state.text === '') {
      await this.delete()
    } else {
      await this.props.update(this.props.id, this.state.text)
    }
    this.setState({
      editing: false,
      unsavedChanges: false
    })
  }

  delete = () =>
    this.props.delete(this.props.id)

  render () {
    const {
      updating,
      removing
    } = this.props

    const inner = this.state.editing
      ? <form onSubmit={
          e => {
            e.preventDefault()
            if (this.state.text === '') return
            this.add()
          }
        }>

          <div className="field">
            <div className="control">
              <textarea
                autoFocus
                onFocus={ this.handleFocus }
                onChange={ this.handleInputChange }
                onKeyPress={ (e) => {
                  if (this.state.text === '') return
                  if (e.shiftKey && e.key === 'Enter') {
                    e.preventDefault()
                    this.add()
                  }
                }}
                defaultValue={ this.state.text }
                className="my-textarea"
                placeholder="Add you text here"
                name="text"/>
            </div>
          </div>
          <section className="listitem-button-container">
            <div className="field is-grouped">
              <p className="control">
                <button
                  onClick={ this.update }
                  className={ 'button ' + (updating ? 'is-loading' : '')}
                  type="button"
                  name="save">
                  Save
                </button>
              </p>
              <p className="control">
                <input
                  onClick={ this.cancel }
                  className="button"
                  type="button"
                  name="cancel edit"
                  value="âœ•"
                />
              </p>
            </div>
            <div className="field is-grouped">
              <p className="control">
                <button
                  onClick={ this.delete }
                  className={ 'button ' + (removing ? 'is-loading' : '')}
                  type="button"
                  name="delete">
                  delete
                </button>
              </p>
            </div>
          </section>
        </form>
      : <p className="listitem-text">
          { this.state.text }
        </p>

    return (
      <li ref={element => { this.element = element }}>
        <section
          className="cursor--pointer"
          onClick={this.enableEditing}>
        { inner }
        </section>
        {
          this.state.unsavedChanges &&
          <section className="unsaved-changes">
            <span className="is-italic">Unsaved Changes</span>
            <section>
              <a onClick={ this.enableEditing }
                className="button is-link">
                View Edits
              </a>
              <a onClick={ this.discardChanges }
                className="button is-link">
                Discard
              </a>
            </section>
          </section>
        }
      </li>
    )
  }
}

export default ListItem
