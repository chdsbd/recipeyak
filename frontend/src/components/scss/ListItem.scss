.listitem-button-container {
  display: flex;
  justify-content: space-between;
}

.listitem-text {
  white-space: pre-wrap;
  line-height: 1.25rem;
  padding-bottom: 0.25rem;
  .on-hover {
    display: none;
  }
  &:hover {
    .on-hover {
      display: block;
    }
  }
}
