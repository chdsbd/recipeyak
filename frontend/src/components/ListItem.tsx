import React from "react"
import Textarea from "react-textarea-autosize"
import { IRecipe } from "@/store/reducers/recipes"
import GlobalEvent from "@/components/GlobalEvent"
import { Markdown } from "@/components/Markdown"
import { Button, ButtonLink } from "@/components/Buttons"
import { hasSelection } from "@/utils/general"
import { normalizeUnitsFracs } from "@/text"

interface IListItemProps {
  readonly id: number
  readonly text?: string
  readonly recipeID?: IRecipe["id"]
  readonly delete: (id: IRecipe["id"]) => void
  readonly update: (
    recipeID: IRecipe["id"],
    id: number,
    data: { text: string },
  ) => void
  readonly removing?: boolean
  readonly updating?: boolean
}

interface IListItemState {
  readonly text: string
  readonly editing: boolean
  readonly unsavedChanges: boolean
}

export default class ListItem extends React.Component<
  IListItemProps,
  IListItemState
> {
  constructor(props: IListItemProps) {
    super(props)
    this.state = {
      text: props.text || "",
      editing: false,
      unsavedChanges: false,
    }
  }

  element = React.createRef<HTMLDivElement>()

  static defaultProps = {
    recipeID: -1,
    removing: false,
  }

  handleKeyDown = (e: KeyboardEvent) => {
    if (e.key === "Escape") {
      this.cancel()
    }
  }

  // ensures that the list item closes when the user clicks outside of the item
  handleGeneralClick = (e: MouseEvent) => {
    /* eslint-disable-next-line @typescript-eslint/consistent-type-assertions */
    const target = e.target as HTMLElement | null
    const el = this.element.current

    if (el == null || target == null) {
      return
    }

    const clickedInComponent = el.contains(target)
    if (clickedInComponent) {
      return
    }
    this.setState((prevState, { text }) => ({
      editing: false,
      unsavedChanges:
        (prevState.editing && prevState.text !== text) ||
        prevState.unsavedChanges,
    }))
  }

  handleInputChange = (e: React.ChangeEvent<HTMLTextAreaElement>) => {
    /* eslint-disable @typescript-eslint/consistent-type-assertions */
    this.setState(({
      [e.target.name]: e.target.value,
    } as unknown) as IListItemState)
    /* eslint-enable @typescript-eslint/consistent-type-assertions */
  }

  enableEditing = () => {
    if (hasSelection()) {
      return
    }

    this.setState({
      editing: true,
      unsavedChanges: false,
    })
  }

  discardChanges = () => {
    this.setState((_, props) => ({
      editing: false,
      text: props.text || "",
      unsavedChanges: false,
    }))
  }

  handleFocus = (e: React.FocusEvent<HTMLTextAreaElement>) => {
    e.target.select()
  }

  handleButtonClick = (e: React.MouseEvent) => {
    e.stopPropagation()
    this.cancel()
  }

  cancel = () => {
    this.setState((_, props) => ({
      editing: false,
      text: props.text || "",
    }))
  }

  update = async (e: React.KeyboardEvent | React.MouseEvent) => {
    e.stopPropagation()
    // if the text is empty, we should just delete the item instead of updating
    if (this.state.text === "") {
      this.delete()
    } else if (this.props.recipeID) {
      this.props.update(this.props.recipeID, this.props.id, {
        text: this.state.text,
      })
    } else {
      this.props.update(-1, this.props.id, {
        text: this.state.text,
      })
    }
    this.setState({
      editing: false,
      unsavedChanges: false,
    })
  }

  delete = () => this.props.delete(this.props.id)

  render() {
    const { updating, removing } = this.props

    const inner = this.state.editing ? (
      <form>
        <GlobalEvent
          mouseUp={this.handleGeneralClick}
          keyDown={this.handleKeyDown}
        />
        <div className="field">
          <div className="control">
            <Textarea
              autoFocus
              onFocus={this.handleFocus}
              onChange={this.handleInputChange}
              onKeyPress={e => {
                if (this.state.text === "") {
                  return
                }
                if (e.shiftKey && e.key === "Enter") {
                  e.preventDefault()
                  this.update(e)
                }
              }}
              defaultValue={this.state.text}
              className="my-textarea"
              placeholder="Add you text here"
              name="text"
            />
          </div>
        </div>
        <section className="listitem-button-container">
          <div className="field is-grouped">
            <p className="control">
              <Button
                size="small"
                onClick={this.update}
                loading={updating}
                name="save">
                Save
              </Button>
            </p>
            <p className="control">
              <Button
                size="small"
                name="cancel edit"
                onClick={this.handleButtonClick}>
                Cancel
              </Button>
            </p>
          </div>
          <div className="field is-grouped">
            <p className="control">
              <Button
                onClick={this.delete}
                size="small"
                loading={removing}
                type="button"
                name="delete">
                Delete
              </Button>
            </p>
          </div>
        </section>
      </form>
    ) : (
      <Markdown>{normalizeUnitsFracs(this.state.text)}</Markdown>
    )

    return (
      <div ref={this.element}>
        <section
          className="cursor-pointer"
          title="click to edit"
          onClick={this.enableEditing}>
          {inner}
        </section>
        {this.state.unsavedChanges && (
          <section className="d-flex justify-space-between align-center">
            <span className="is-italic">Unsaved Changes</span>
            <section>
              <ButtonLink size="small" onClick={this.enableEditing}>
                View Edits
              </ButtonLink>
              <ButtonLink size="small" onClick={this.discardChanges}>
                Discard
              </ButtonLink>
            </section>
          </section>
        )}
      </div>
    )
  }
}
