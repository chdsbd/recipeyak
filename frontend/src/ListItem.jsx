import React from 'react'
import PropTypes from 'prop-types'

import './ListItem.scss'

class ListItem extends React.Component {
  constructor (props) {
    super(props)
    this.state = {
      text: props.text || '',
      editing: false,
      unsavedChanges: false,
    }

    document.addEventListener('mousedown', (e) => this.handleGeneralClick(e))
  }

  componentWillUnmount () {
    document.removeEventListener('mousedown', (e) => this.handleGeneralClick(e))
  }

  // ensures that the list item closes when the user clicks outside of the item
  handleGeneralClick (e) {
    e.stopPropagation()
    const el = this.refs.listitem
    if (el == null) return
    const clickOnListItem = el.contains(e.srcElement)
    if (!clickOnListItem) {
      this.setState((prevState, props) => ({
        editing: false,
        unsavedChanges: (prevState.editing && prevState.text !== props.text) || prevState.unsavedChanges,
      }))
    }
  }

  handleInputChange (e) {
    this.setState({ [e.target.name]: e.target.value })
  }

  enableEditing () {
    this.setState({
      editing: true,
      unsavedChanges: false,
    })
  }

  discardChanges () {
    this.setState((prevState, props) => ({
      editing: false,
      text: props.text,
      unsavedChanges: false,
    }))
  }

  handleFocus (event) {
    event.target.select()
  }

  add () {
    console.log(this.state)
    // TODO: update store
  }

  cancel (e) {
    e.stopPropagation()
    this.setState((prevState, props) => ({
      editing: false,
      text: props.text,
    }))
  }

  update (e) {
    e.stopPropagation()
    this.setState({
      editing: false,
      unsavedChanges: false,
    })
    // if the text is empty, we should just delete the item instead of updating
    if (this.state.text === '') {
      this.props.delete(this.props.index)
    } else {
      this.props.update(this.props.index, this.state.text)
    }
  }

  delete (index) {
    this.props.delete(index)
  }

  render () {
    const inner = this.state.editing
      ? <form onSubmit={
          (e) => {
            e.preventDefault()
            if (this.state.text === '') return
            this.add()
          }
        }>

          <div className="field">
            <div className="control">
              <textarea
                autoFocus
                onFocus={ (e) => this.handleFocus(e) }
                onChange={ (e) => this.handleInputChange(e) }
                onKeyPress={ (e) => {
                  if (this.state.text === '') return
                  if (e.shiftKey && e.key === 'Enter') {
                    e.preventDefault()
                    this.add()
                  }
                }}
                defaultValue={ this.state.text }
                className="textarea"
                placeholder="Add you text here"
                name="text"/>
            </div>
          </div>
          <section className="listitem-button-container">
            <div className="field is-grouped">
              <p className="control">
                <input
                  onClick={ e => this.update(e) }
                  className="button"
                  type="button"
                  name="save"
                  value="Save"
                />
              </p>
              <p className="control">
                <input
                  onClick={ e => this.cancel(e) }
                  className="button"
                  type="button"
                  name="cancel edit"
                  value="âœ•"
                />
              </p>
            </div>
            <div className="field is-grouped">
              <p className="control">
                <input
                  onClick={ () => this.delete(this.props.index) }
                  className="button"
                  type="button"
                  name="delete"
                  value="delete"
                />
              </p>
            </div>
          </section>
        </form>
      : <p className="listitem-text">
          { this.state.text }
        </p>

    return (
      <li ref="listitem">
        <section
          className="cursor--pointer"
          onClick={() => this.enableEditing()}>
        { inner }
        </section>
        {
          this.state.unsavedChanges &&
          <section className="unsaved-changes">
            <span className="is-italic">Unsaved Changes</span>
            <section>
              <a onClick={() => this.enableEditing() }
                className="button is-link">
                View Edits
              </a>
              <a onClick={() => this.discardChanges() }
                className="button is-link">
                Discard
              </a>
            </section>
          </section>
        }
      </li>
    )
  }
}

ListItem.PropTypes = {
  text: PropTypes.string,
  index: PropTypes.number,
  delete: PropTypes.func,
  update: PropTypes.func,
}

export default ListItem
