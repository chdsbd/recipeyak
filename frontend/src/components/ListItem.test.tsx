import React from "react"

import { mount, configure } from "enzyme"
import Adapter from "enzyme-adapter-react-16"

import { MemoryRouter } from "react-router"

import ListItem from "./ListItem"

configure({ adapter: new Adapter() })

describe("<ListItem/>", () => {
  it("renders without crashing", () => {
    mount(
      <MemoryRouter>
        <ListItem id={0} delete={jest.fn()} update={jest.fn()} />
      </MemoryRouter>
    )
  })
})
