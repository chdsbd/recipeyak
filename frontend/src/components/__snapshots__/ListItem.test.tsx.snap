// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListItem/> snap with capitalization 1`] = `
<div>
  <section
    className="cursor-pointer"
    onClick={[Function]}
    title="click to edit"
  >
    <p>
      Add 3 Tablespoons of tomato paste. Cook 1 minute.
    </p>
  </section>
</div>
`;

exports[`<ListItem/> snap with capitalization 2`] = `
<div>
  <section
    className="cursor-pointer"
    onClick={[Function]}
    title="click to edit"
  >
    <p>
      Measure 1 teaspoon of rice vinegar.
    </p>
  </section>
</div>
`;

exports[`<ListItem/> snap with capitalization 3`] = `
<div>
  <section
    className="cursor-pointer"
    onClick={[Function]}
    title="click to edit"
  >
    <p>
      Add 1 Tablespoon + 2 teaspoons of soy sauce.
    </p>
  </section>
</div>
`;

exports[`<ListItem/> snap with capitalization 4`] = `
<div>
  <section
    className="cursor-pointer"
    onClick={[Function]}
    title="click to edit"
  >
    <p>
      Stir in 1 teaspoon salt
    </p>
  </section>
</div>
`;
